extensions/rich_text_components/Video/Video.js

```diff
@@ -25,19 +25,55 @@ oppia.directive('oppiaNoninteractiveVideo', [
       restrict: 'E',
       scope: {},
       templateUrl: 'richTextComponent/Video',
-      controller: ['$scope', '$attrs', 'explorationContextService',
-        function($scope, $attrs, explorationContextService) {
+      controller: ['$scope', '$attrs', 'explorationContextService', '$element',
+        'autoplayedVideosService', 'PAGE_CONTEXT', '$timeout',
+        function($scope, $attrs, explorationContextService, $element,
+          autoplayedVideosService, PAGE_CONTEXT, $timeout) {
           var start = oppiaHtmlEscaper.escapedJsonToObj($attrs.startWithValue);
           var end = oppiaHtmlEscaper.escapedJsonToObj($attrs.endWithValue);
 
-          $scope.autoplaySuffix = (oppiaHtmlEscaper.escapedJsonToObj(
-            $attrs.autoplayWithValue) ? '&autoplay=1' : '&autoplay=0');
           $scope.videoId = oppiaHtmlEscaper.escapedJsonToObj(
             $attrs.videoIdWithValue);
           $scope.timingParams = '&start=' + start + '&end=' + end;
-          $scope.videoUrl = $sce.trustAsResourceUrl(
-            'https://www.youtube.com/embed/' + $scope.videoId + '?rel=0' +
-            $scope.timingParams + $scope.autoplaySuffix);
+          $scope.autoplaySuffix = '&autoplay=0';
+
+          $timeout(function() {
+            // Check whether creator wants to autoplay this video or not
+            var autoplayVal = oppiaHtmlEscaper.escapedJsonToObj(
+              $attrs.autoplayWithValue);
+
+            // This code helps in visibility of video. It checks whether
+            // mid point of video frame is in the view or not.
+            var rect = angular.element($element)[0].getBoundingClientRect();
+            var clh = window.innerHeight;
+            var clw = window.innerWidth;
+            var isVisible = ((rect.left + rect.right) / 2 < clw &&
+              (rect.top + rect.bottom) / 2 < clh) &&
+              (rect.left > 0 && rect.right > 0);
+
+            // Autoplay if user is in lerner view and creator has specified
+            // to autoplay given video.
+            if (explorationContextService.getPageContext() ===
+              PAGE_CONTEXT.LEARNER && autoplayVal) {
+              // If it has been autoplayed then do not autoplay again.
+              if (
```

Comment:
Actually, you've already set autoplaySuffix outside this, so really you only need the following (right?) instead of lines 59-67:

```
    if (!autoplayedVideosService.hasVideoBeenAutoplayed($scope.videoId) && isVisible) {
      $scope.autoplaySuffix = '&autoplay=1';
      autoplayedVideosService.addAutoplayedVideo($scope.videoId);
    }
```